packages/server/lib/browsers/cdp_automation.ts

```diff
@@ -0,0 +1,130 @@
+import _ from 'lodash'
+import cdp from 'devtools-protocol'
+import tough from 'tough-cookie'
+
+interface CyCookie {
+  name: string
+  value: string
+  expirationDate: number
+  hostOnly: boolean
+  domain: string
+  path: string
+  secure: boolean
+  httpOnly: boolean
+}
+
+const cors = require('../util/cors')
+
+export function CdpAutomation (opts: {
+  invokeViaDebugger: (message: string, data?: any) => Promise<any>
+}) {
+  const { invokeViaDebugger } = opts
+
+  const normalizeGetCookieProps = function (cookie: cdp.Network.Cookie): CyCookie {
+    if (cookie.expires === -1) {
+      delete cookie.expires
+    }
+
+    // @ts-ignore
+    cookie.expirationDate = cookie.expires
+    delete cookie.expires
+
+    // @ts-ignore
+    return cookie
+  }
+
+  const normalizeGetCookies = function (cookies: cdp.Network.Cookie[]) {
+    return _.map(cookies, normalizeGetCookieProps)
+  }
+
+  const normalizeSetCookieProps = function (cookie: CyCookie): cdp.Network.SetCookieRequest {
+    cookie.name || (cookie.name = '') // name can't be undefined/null
+    cookie.value || (cookie.value = '') // ditto
+    // @ts-ignore
+    cookie.expires = cookie.expirationDate
+    if (!cookie.hostOnly && cookie.domain[0] !== '.') {
+      let parsedDomain = cors.parseDomain(cookie.domain)
+
+      // not a top-level domain (localhost, ...) or IP address
+      if (parsedDomain && parsedDomain.tld !== cookie.domain) {
+        cookie.domain = `.${cookie.domain}`
+      }
+    }
+
+    delete cookie.hostOnly
+    delete cookie.expirationDate
+
+    return cookie
+  }
+
+  const getAllCookies = function (data) {
+    return invokeViaDebugger('Network.getAllCookies').then(function (result: cdp.Network.GetAllCookiesResponse) {
+      return normalizeGetCookies(result.cookies).filter(function (cookie: CyCookie) {
+        return _.every([
+          !data.domain || tough.domainMatch(cookie.domain, data.domain),
+          !data.path || tough.pathMatch(cookie.path, data.path),
+          !data.name || data.name === cookie.name,
```

Comment:
this is an extremely dense function that likely needs to be extracted into a method that describes what this is doing

Reply:
b2825a693